x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/TransportPutFollowAction.java

```diff
@@ -128,93 +131,123 @@ private void createFollowerIndex(
-                // Overwriting UUID here, because otherwise we can't follow indices in the same cluster
-                settingsBuilder.put(IndexMetaData.SETTING_INDEX_UUID, UUIDs.randomBase64UUID());
-                settingsBuilder.put(IndexMetaData.SETTING_INDEX_PROVIDED_NAME, followIndex);
-                settingsBuilder.put(CcrSettings.CCR_FOLLOWING_INDEX_SETTING.getKey(), true);
-                settingsBuilder.put(IndexSettings.INDEX_SOFT_DELETES_SETTING.getKey(), true);
-                imdBuilder.settings(settingsBuilder);
-
-                // Copy mappings from leader IMD to follow IMD
-                for (ObjectObjectCursor<String, MappingMetaData> cursor : leaderIndexMetaData.getMappings()) {
-                    imdBuilder.putMapping(cursor.value);
+                @Override
+                public void onFailure(Exception e) {
+                    listener.onFailure(e);
                 }
-                imdBuilder.setRoutingNumShards(leaderIndexMetaData.getRoutingNumShards());
-                IndexMetaData followIMD = imdBuilder.build();
-                mdBuilder.put(followIMD, false);
+            });
+    }
 
-                ClusterState.Builder builder = ClusterState.builder(currentState);
-                builder.metaData(mdBuilder.build());
-                ClusterState updatedState = builder.build();
+    private void initiateRestore(RestoreService.RestoreRequest restoreRequest, ActionListener<RestoreSnapshotResponse> listener) {
+        threadPool.executor(ThreadPool.Names.SNAPSHOT).execute(() ->
+            restoreService.restoreSnapshot(restoreRequest, new ActionListener<RestoreService.RestoreCompletionResponse>() {
+                @Override
+                public void onResponse(RestoreService.RestoreCompletionResponse restoreCompletionResponse) {
+                    final Snapshot snapshot = restoreCompletionResponse.getSnapshot();
 
-                RoutingTable.Builder routingTableBuilder = RoutingTable.builder(updatedState.routingTable())
-                        .addAsNew(updatedState.metaData().index(request.getFollowRequest().getFollowerIndex()));
-                updatedState = allocationService.reroute(
-                        ClusterState.builder(updatedState).routingTable(routingTableBuilder.build()).build(),
-                        "follow index [" + request.getFollowRequest().getFollowerIndex() + "] created");
+                    ClusterStateListener clusterStateListener = new ClusterStateListener() {
+                        @Override
+                        public void clusterChanged(ClusterChangedEvent changedEvent) {
+                            RestoreInProgress.Entry prevEntry = RestoreService.restoreInProgress(changedEvent.previousState(), snapshot);
+                            RestoreInProgress.Entry newEntry = RestoreService.restoreInProgress(changedEvent.state(), snapshot);
+                            if (prevEntry == null) {
```

Comment:
this block of code is copy-pasted from TransportRestoreSnapshotAction. Do we need this here? Can we possibly share code?